---
created: 2024-06-04 20:58
updated: 2025-06-10 09:54
dg-publish: true
tags:
  - garden/ğŸŒ³Evergreen
---

## 00.Preparing

### 00.Notice
- **ä¿ç•™æ‰€æœ‰çš„è„šæœ¬ã€æäº¤ä»»åŠ¡çš„å‘½ä»¤ã€æ¯ä¸ªæ­¥éª¤çš„æ—¥å¿—æ–‡ä»¶ã€‚**
- **é‡‡ç”¨ä¸‹æ–‡æŒ‡å®šçš„æµç¨‹ã€è½¯ä»¶ç‰ˆæœ¬å’Œå‚æ•°ï¼Œæ¯ä¸ªæ­¥éª¤çš„è¾“å‡ºæ–‡ä»¶ååç¼€å¿…é¡»åŒç¤ºä¾‹ä¿æŒä¸€è‡´ã€‚**
- **ç¡®ä¿è½¯ä»¶æ‰€åœ¨ç›®å½•å·²å­˜åœ¨äºPATHå˜é‡ä¸­ï¼Œæˆ–è‡ªè¡Œæ·»åŠ ç»å¯¹è·¯å¾„ã€‚**

### 01.Software Version
- bwa==0.7.17
- gatk==4.4.0.0
- samtools==1.18
- bcftools==1.18
- fastp==0.23.4

### 02.Creating Directories

```shell
mkdir log 00.reference 01.fastq 02.bam 03.gvcf 04.combined_gvcf 05.combined_vcf 06.filtered_vcf 07.concatenate
```

### 03.Reference Genome

0. ä»¥NCBIä¸Š*Bos taurus*æœ€æ–°çš„ARS-UCD2.0ç‰ˆæœ¬åŸºå› ç»„ä¸ºå‚è€ƒåŸºå› ç»„ï¼Œå¹¶ä¿å­˜è‡³ç›®å½•`00.reference`ä¸­,å¯¹æŸ“è‰²ä½“è¿›è¡Œé‡å‘½åï¼ˆå¸¸æŸ“è‰²ä½“ï¼ŒXï¼ŒTï¼ŒMTé‡å‘½åï¼Œscafoldsä¸ç”¨ç®¡ï¼‰ï¼Œpythonè„šæœ¬åœ¨scriptsç›®å½•ä¸‹ï¼š

```
wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/002/263/795/GCF_002263795.3_ARS-UCD2.0/GCF_002263795.3_ARS-UCD2.0_genomic.fna.gz
python3 scripts/ChangeChromosomeNameInFastaOrGff.py -i GCF_002263795.3_ARS-UCD2.0_genomic.fna.gz --fa -g -N chr_name -o GCF_002263795.3_ARS-UCD2.0_genomic.rename.fna
```

é€šè¿‡ä»¥ä¸‹æ­¥éª¤ä¸ºå‚è€ƒåŸºå› ç»„å»ºç«‹ç´¢å¼•ï¼š

1. Generate the BWA index using BWA. This will create the â€œ.fasta.ambâ€, â€œ.fasta.annâ€, â€œ.fasta.bwtâ€, â€œ.fasta.pacâ€ and â€œ.fasta.saâ€ files.
```shell
bwa index GCF_002263795.1_ARS-UCD1.2_genomic.addY.rename.fna
```
Â 
2. Generate the FASTA file index using samtools. This will create the â€œ.fasta.faiâ€ file.
```shell
samtools faidx GCF_002263795.1_ARS-UCD1.2_genomic.addY.rename.fna
```
Â 
3. Generate the sequence dictionary using Picard. This will create the â€œ.dictâ€ file.
```shell
java -jar picard_2.18.17.jar CreateSequenceDictionary R=GCF_002263795.1_ARS-UCD1.2_genomic.addY.rename.fna O=GCF_002263795.1_ARS-UCD1.2_genomic.addY.rename.dict
```


> **NOTEï¼šå¼€å§‹ä¹‹å‰ï¼Œæ£€æŸ¥æ‰€æœ‰çš„æ ·æœ¬çš„fastqæ–‡ä»¶ï¼Œæ˜¯å¦éƒ½æ˜¯1å¯¹ï¼Œå¦‚æœæŸäº›æ ·æœ¬å­˜åœ¨å¤šä¸ªlaneå¯¼è‡´ä¸‹æœºæ•°æ®ä¸æ­¢1å¯¹fastqæ–‡ä»¶ï¼Œç›´æ¥æŠŠè¯¥æ ·æœ¬æ‰€æœ‰çš„fastq1åˆå¹¶åœ¨ä¸€èµ·ï¼Œæ‰€æœ‰çš„fastq2åˆå¹¶åœ¨ä¸€èµ·ï¼Œå†è¿›è¡Œæ¥ä¸‹æ¥çš„æ“ä½œã€‚** 



## 01.Quality Control

é‡‡ç”¨`fastp`è¿›è¡Œè´¨æ§ï¼Œé»˜è®¤å‚æ•°å³å¯ï¼Œä¿ç•™`report`ï¼š
ä¸‹é¢ä»£ç æ¡†å†…å®¹ä¿å­˜ä¸º`01.cleaning.sh`:

```shell
# 01.cleaning.sh
${sm}=$1
fastp -i ${sm}_1.fq.gz -I ${sm}_2.fq.gz -o ${sm}_1.clean.fq.gz -O ${sm}_2.clean.fq.gz -h ${sm}.html -j ${sm}.json -w 8
```

- æ— è®ºraw.fq.gzå‰ç¼€æ˜¯ä»€ä¹ˆæ ¼å¼ï¼Œclean.fq.gzç»Ÿä¸€å‘½åä¸ºsample_1.clean.fq.gzåŠsample_2.clean.fq.gz
- -w çº¿ç¨‹æ•°è‡ªè¡Œè°ƒèŠ‚

é€šè¿‡ä»¥ä¸‹å‘½ä»¤å°†ä¸Šé¢è„šæœ¬æäº¤ä»»åŠ¡è‡³è®¡ç®—èŠ‚ç‚¹ï¼ˆä»¥slurmä¸ºä¾‹ï¼‰ï¼Œæ¯ä¸ªæ ·æœ¬ä¸€ä¸ªä»»åŠ¡ï¼ˆåˆç†è°ƒæ•´çº¿ç¨‹å’Œå†…å­˜ï¼‰ã€‚ï¼š

```
# æ‰€æœ‰æ ·æœ¬åä¿å­˜è‡³æ–‡ä»¶lstä¸­ï¼Œæ¯è¡Œ1ä¸ª
for i in `cat lst`; do sbatch -J ${sm} -p normal -N 1 -c 32 -o log/${sm}_cleaning.o -e log/${sm}_cleaning.e 01.cleaning.sh ${sm}; done
```

## 02.Mapping

å¼€å§‹ä¹‹å‰ï¼Œéœ€è¦å‡†å¤‡æ ·æœ¬listæ–‡ä»¶ï¼Œä¸‰åˆ—åˆ†åˆ«æ˜¯ï¼šæ ·æœ¬åï¼Œfq1çš„ç»å¯¹è·¯å¾„ï¼Œfq2çš„ç»å¯¹è·¯å¾„ï¼Œä»¥tabåˆ†å‰²ã€‚
**ä¸è¦æ‰€æœ‰æ ·æœ¬çš„bamæ”¾åœ¨åŒä¸€ä¸ªæ–‡ä»¶å¤¹ä¸‹é¢ï¼Œè¦ä¸ºæ¯ä¸ªæ ·æœ¬åˆ›å»ºå•ç‹¬çš„ç›®å½•ï¼š**

```shell
cd 02.mapping
for i in `cut -f1 sample_lst`; do mkdir ${i}; done
cd ..
```

ä¸‹é¢ä»£ç æ¡†å†…å®¹ä¿å­˜ä¸º`02.mapping.sh`:

```shell
# 02.mapping.sh
# define variable
export sm=${1} #sample nameï¼Œä½œä¸ºè¯¥è„šæœ¬å¤–éƒ¨çš„ç¬¬1ä¸ªå‚æ•°
export fq1=${2} #fq1çš„ç»å¯¹è·¯å¾„ï¼Œä½œä¸ºè¯¥è„šæœ¬å¤–éƒ¨çš„ç¬¬2ä¸ªå‚æ•°
export fq2=${3} #fq2çš„ç»å¯¹è·¯å¾„ï¼Œä½œä¸ºè¯¥è„šæœ¬å¤–éƒ¨çš„ç¬¬3ä¸ªå‚æ•°
export thread=8 #çº¿ç¨‹æ•°ï¼Œæ ¹æ®é›†ç¾¤æƒ…å†µè‡ªè¡Œä¿®æ”¹
export ref=/path/00.reference/GCF_002263795.1_ARS-UCD1.2_genomic.addY.rename.fna  #å‚è€ƒåŸºå› ç»„,è®°å¾—æ”¹æˆç»å¯¹è·¯å¾„
Â 
# mapping and sorting

bwa mem -t ${thread} -R '@RG\tID:'${sm}'\tSM:'${sm}'\tLB:'${sm}'\tPL:ILLUMINA' ${sm} ${ref} ${fq1} ${fq2} | samtools sort -o 02.bam/${sm}/${sm}.sorted.bam - -@ ${thread}

# dedup
gatk MarkDuplicates --spark-runner LOCAL -I 02.bam/${sm}/${sm}.sorted.bam -O 02.bam/${sm}/${sm}.dedup.bam -M 02.bam/${sm}/${sm}.dup.metrics --CREATE_INDEX true --VALIDATION_STRINGENCY SILENT --REMOVE_DUPLICATES true

#stat
samtools flagstat 02.bam/${sm}/${sm}.dedup.bam > 02.bam/${sm}/${sm}.dedup.flagstat

samtools coverage 02.bam/${sm}/${sm}.dedup.bam > 02.bam/${sm}/${sm}.dedup.coverage

```
é€šè¿‡ä»¥ä¸‹å‘½ä»¤å°†ä¸Šé¢è„šæœ¬æäº¤ä»»åŠ¡è‡³è®¡ç®—èŠ‚ç‚¹ï¼ˆä»¥slurmä¸ºä¾‹ï¼‰ï¼Œæ¯ä¸ªæ ·æœ¬ä¸€ä¸ªä»»åŠ¡ï¼ˆåˆç†è°ƒæ•´çº¿ç¨‹å’Œå†…å­˜ï¼‰ï¼š

```
cat sample_lst | while read sm fq1 fq2; do sbatch -J ${sm} -p normal -N 1 -c 32 -o log/${sm}_mapping.o -e log/${sm}_mapping.e 02.mapping.sh ${sm} ${fq1} ${fq2}; done
```



## 03.Calling

### 01.HaplotypeCaller

å¼€å§‹ä¹‹å‰ï¼Œéœ€è¦æå‰å‡†å¤‡åŒ…å«æ ·æœ¬åå’Œbamæ–‡ä»¶ç»å¯¹è·¯å¾„çš„listæ–‡ä»¶ï¼Œä¸€å…±2åˆ—ï¼Œä»¥tabåˆ†å‰²ã€‚
**ä¸è¦æ‰€æœ‰æ ·æœ¬çš„gvcfæ”¾åœ¨åŒä¸€ä¸ªæ–‡ä»¶å¤¹ä¸‹é¢ï¼Œè¦ä¸ºæ¯ä¸ªæ ·æœ¬åˆ›å»ºå•ç‹¬çš„ç›®å½•ï¼š**

```shell
cd 03.gvcf
for i in `cut -f1 bam_lst`; do mkdir ${i}; done
cd ..
```

ä¸‹é¢ä»£ç æ¡†å†…å®¹ä¿å­˜ä¸º`03.HaplotypeCaller.sh`:
```shell
# 03.HaplotypeCaller.sh
# HaplotypeCaller
export sm=${1}
export bam=${2}
export chr=${3}
export ref=/path/00.reference/GCF_002263795.1_ARS-UCD1.2_genomic.addY.rename.fna
Â 
gatk HaplotypeCaller -R ${ref} -I ${bam} -L ${chr} -ERC GVCF -O 03.gvcf/${sm}/${sm}.chr${chr}.gvcf.gz
```

è¿™ä¸€æ­¥è¦æ±‚åˆ†æŸ“è‰²ä½“è¿›è¡Œcallingï¼Œ**éœ€è¦29æ¡å¸¸æŸ“è‰²ä½“åŠ ä¸ŠX Y MTå…±32æ¡æŸ“è‰²ä½“**ï¼Œæ¯ä¸ªä¸ªä½“æ¯æ¡æŸ“è‰²ä½“åˆ†åˆ«é€’äº¤ä»»åŠ¡ï¼Œæ³¨æ„åµŒå¥—å¾ªç¯ï¼š
```
cat bam_lst | while read sm bam; do for chr in {1..29} X Y MT; do sbatch -J ${sm} -p normal -N 1 -c 32 -o log/${sm}_${chr}_calling.o -e log/${sm}_${chr}_calling.e 03.HaplotypeCaller.sh ${sm} ${bam} ${chr}; done; done
```


---


### 02.CombineGVCFs and GenotypeGVCFs
ä¸‹é¢ä»£ç æ¡†å†…å®¹ä¿å­˜ä¸º`04.05.Combine_Genotype_GVCFs.sh`:
```shell
# 04.05.Combine_Genotype_GVCFs.sh
export chr=${1}
export ref=/path/00.reference/GCF_002263795.1_ARS-UCD1.2_genomic.addY.rename.fna

# CombineGVCFs 
find 03.gvcf/ -name "*.chr${chr}.gvcf.gz" > 04.combined_gvcf/chr${chr}.gvcf.lst
gatk CombineGVCFs -R ${ref} -V 04.combined_gvcf/chr${chr}.gvcf.lst -O 04.combined_gvcf/chr${chr}.gvcf.gz
Â 
# GenotypeGVCFs
gatk GenotypeGVCFs -R ${ref} -V 04.combined_gvcf/chr${chr}.gvcf.gz -O 05.combined_vcf/chr${chr}.vcf.gz
```

æŒ‰ç…§ä¸Šé¢çš„ç¤ºä¾‹ï¼Œæ¯æ¡æŸ“è‰²ä½“æäº¤ä¸€æ¬¡ä»»åŠ¡ï¼š
```
for i in {1..29} X Y MT; do 
```


### 03.filtering
ä¸‹é¢ä»£ç æ¡†å†…å®¹ä¿å­˜ä¸º`06.filtering.sh`:
```shell
# 06.filtering.sh

export chr=${1}

# HardFilter
bcftools view \
-v snps \
-e 'QD < 2.0 || QUAL < 30.0 || SOR > 3.0 || FS > 60.0 || MQ < 40.0 || MQRankSum < -12.5 || ReadPosRankSum < -8.0' \
-m2 -M2 \
05.combined_vcf/chr${chr}.vcf.gz \
-Oz \
-o 06.filtered_vcf/chr${chr}.filter.vcf.gz

# think over the threshold
bcftools view -i 'F_MISSING < 0.1 & MAF > 0.05' 06.filtered_vcf/chr${chr}.filter.vcf.gz -Oz -o 06.filtered_vcf/chr${chr}.filtered.vcf.gz

# index
bcftools index 06.filtered_vcf/chr${chr}.filtered.vcf.gz
```

> The slf subcommand of BaseNumber were used to filter low-quality variants with the parameter recommended by GATK best practices: (1) SNP clusters with â€œ-clusterSize 3â€ and â€œ-clusterWindowSize 10â€ options; (2) SNPs with mean depth (for all samples) < 1/3Ã— and > 3Ã— (Ã—, overall mean sequencing depth across all SNPs); (3) quality by depth, QD < 2; (4) phred-scaled variant quality score, QUAL < 30; (5) strand odds ratio, SOR > 3; (6) Fisher strand, FS > 60; (7) mapping quality, MQ < 40; (8) mapping quality rank sum test, MQRankSum <âˆ’12.5; and (9) read position rank sum test, ReadPosRankSum < âˆ’8.



### 04.concatenate

å°†å…¨éƒ¨ä¸ªä½“çš„æ¯æ¡æŸ“è‰²ä½“åˆå¹¶åœ¨ä¸€èµ·ï¼Œç”Ÿæˆå…¨éƒ¨æ ·æœ¬å…¨åŸºå› ç»„çš„vcfï¼š
ä¸‹é¢ä»£ç æ¡†å†…å®¹ä¿å­˜ä¸º`07.concatenate.sh`:

```shell
# vcf lst
find 06.filtered_vcf/ -name chr*.filtered.vcf.gz | sort -V > 07.concatenate/vcf_lst

# concat
bcftools concat -f 07.concatenate/vcf_lst -Oz -o 07.concatenate/all_samples.all_chr.vcf.gz

# index 
bcftools index 07.concatenate/all_samples.all_chr.vcf.gz
```

